### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by Aliblas**
This contains all the starting code for authentication.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.sps.util.UserHelper;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+    // Request: param redirectUrl for after login is finished
```
Comment (isamitre) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: the indentation should be two spaces, not a tab
[Java Style Guide](https://google.github.io/styleguide/javaguide.html#s4.2-block-indentation)

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.sps.util.UserHelper;
```
Comment (isamitre) [inline src/main/java/com/google/sps/servlets/LoginServlet.java]: import statements should be alphabetical

File: src/main/java/com/google/sps/servlets/UserStatusServlet.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.sps.util.UserHelper;
```
Comment (isamitre) [inline src/main/java/com/google/sps/servlets/UserStatusServlet.java]: import statements should be alphabetical and in one block
[Java Style Guide](https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing)

File: src/main/java/com/google/sps/servlets/UserStatusServlet.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+@WebServlet("/user-status")
+public class UserStatusServlet extends HttpServlet {
+
+    // Response: returns some user information JSON from the current user's sign-in API profile.
```
Comment (isamitre) [inline src/main/java/com/google/sps/servlets/UserStatusServlet.java]: indentation should be two spaces

File: src/main/java/com/google/sps/util/UserHelper.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package com.google.sps.util;
+
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+
+public class UserHelper {
+
+    // returns: opaque String unique ID of the currently logged-in user.
```
Comment (isamitre) [inline src/main/java/com/google/sps/util/UserHelper.java]: indentation should be two spaces

File: src/main/webapp/javascript/UserAuth.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+
+// To be called before any request that requires user be logged in.
+// Action: if user not logged in, redirects to login page
+// redirects to redirectUrl regardless.
+// Args: String url redirectUrl defaults to the page function was called form.
+function confirmUserAndRedirect(redirectUrl) {
+    if (!redirectUrl) {
```
Comment (isamitre) [inline src/main/webapp/javascript/UserAuth.js]: indentation should be two spaces
[JavaScript Style Guide](https://google.github.io/styleguide/jsguide.html#formatting-block-indentation)

**Review by isamitre — CHANGES_REQUESTED** [review]

**Comment by Aliblas** [comment]
@googlebot I signed it!

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+  // Request: param redirectUrl for after login is finished
+  // Response: if not logged in, redirects to login page -> redirectUrl,
+  //           otherwise redirects to redirectUrl.
+  @Override
+  public void doGet (HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    String redirectUrl = request.getParameter("redirectUrl");
+    if (!userService.isUserLoggedIn()) {
+        String loginPageUrl = userService.createLoginURL(redirectUrl);
```
Comment (isamitre) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:41]: indentation should be two spaces

File: src/main/java/com/google/sps/servlets/UserStatusServlet.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  // Data: "isUserLoggedIn" : boolean
+  @Override
+  public void doGet (HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    UserService userService = UserServiceFactory.getUserService();
+
+    JsonObject userInfo = new JsonObject();
+    userInfo.addProperty("isUserLoggedIn", userService.isUserLoggedIn());
+    if (userInfo.get("isUserLoggedIn").getAsBoolean()) {
+        System.out.println("User is logged in");
```
Comment (isamitre) [inline src/main/java/com/google/sps/servlets/UserStatusServlet.java:43]: indentation should be two spaces and same goes for the else statement

**Review by isamitre — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  // Request: param redirectUrl for after login is finished
+  // Response: if not logged in, redirects to login page -> redirectUrl,
+  //           otherwise redirects to redirectUrl.
+  @Override
+  public void doGet (HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (oturner24) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:37]: nit: remove the space between doGet and (

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    UserService userService = UserServiceFactory.getUserService();
+    String redirectUrl = request.getParameter("redirectUrl");
+    if (!userService.isUserLoggedIn()) {
+        String loginPageUrl = userService.createLoginURL(redirectUrl);
+        response.sendRedirect(loginPageUrl);
+        return;
+    }
+
+    response.sendRedirect(redirectUrl);
+    return;
```
Comment (oturner24) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:47]: You can remove this - there's no need to explicitly return at the end of the method

File: src/main/java/com/google/sps/servlets/UserStatusServlet.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+
+@WebServlet("/user-status")
+public class UserStatusServlet extends HttpServlet {
+
+  // Response: returns some user information JSON from the current user's sign-in API profile.
+  // Data: "isUserLoggedIn" : boolean
+  @Override
+  public void doGet (HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (oturner24) [inline src/main/java/com/google/sps/servlets/UserStatusServlet.java:36]: nit: no space after method name

**Review by oturner24 — APPROVED** [review]

**Pull request merged**